README
#
# py-htbstat README file
# $Revision: 1.5 $
#
# Copyright (c) 2005 Dmytro O. Redchuk
# Copyright (c) 2005 VOLZ Llc
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#



Introduction
============

py-htbstat is a tool set which helps to collect, monitor and analyze htb
classes statistic.

HTB statistic data are captured, processed and saved into RRD databases
along with HTB classes dumps. This statistic data then may be displayed by
CGI scripts and processed further by some checkers.

py-htbstat is a small set of scripts actually. Some part of functionality
was separated into a python module, HTBstat.

py-htbstat requires py-rrdtool (originally i've taken it from
http://www.nongnu.org/py-rrdtool/, but it seems  that now any linux distro
does have an "official" python-to-rrd interface, rrdpython --
http://oss.oetiker.ch/rrdtool/prog/rrdpython.en.html -- in debian it is called
python-rrdtool now).

Below, in "Files" section, there are some descriptions and explanations.

And far below, in "Where to use" section you'll find some additional notes.



Files
=====

bin/
	makestat.sh	-- script to be run by cron, takes
			   `tc -s -d class show ...' output and passes it to
			   statsfill.py.  See etc/cron.htbstat for an example
	statsfill.py	-- takes `tc -s -d class show ...' from makestat.sh
			   and fills RRD bases with stats. Also saves HTB
			   classes (using pickle module) in RRD bases
			   directory. Creates RRD bases on the fly.
			   NOTE, that these two are sample scripts only and
			   doesn't exist in rpm file.
cgi-bin/
	dumpclass.py	-- small CGI script which "dumps" HTB class pickled
			   by statsfill.py to browser screen. Also, if run
			   in box where your classes are configured, dumps
			   required class "in real time".
	htbcheck.py	-- this CGI script checks HTB classes hierarchy and
			   displays warning if some conditions are unmet
			   (from lighter one to the worst one imho):
			   sum of children's ceil higher than parent's ceil,
			   sum of children's rate higher than parent's rate,
			   sum of children's rate higher than parent's ceil.
	htbstat.py	-- 'main window' which allows to select (with
			   multiple selection possibility) classes,
			   timeranges, units, and draws a lot of pictures
			   from data stored in RRD bases. Allows to select
			   by class id or parent class id. For example, you
			   can select classes 33 (x:33 here) 48, 1049 etc.
			   and enjoy pictures for these, or you can view
			   pictures for classes, which parents are 21, 557,
			   2063, 3fa0 etc.
	htbstatpic.py	-- script, which accepts the same parameters, as `htbstat.py'
			   but returns a single picture (for the first class, if
			   multiple requested), can be included as <IMG src=...>
			   in other pages.
doc/
	LICENSE		-- GNU GPL.
	README		-- this file.
etc/
	functions	-- common functions, which are too uncommon to be
			   included into the module :-)
			   "Called" by htbstat.cgi.
	htbstat.cgi	-- config file for CGI scripts. "Called" by
			   htbstat.conf.
	htbstat.conf	-- 'main' config file, contains some kind of
			   documentation, too.
			   Configuration files are python commands lists
			   actually.
	apache2.htbstat -- example Apache2 config
	cron.htbstat -- example cron script
HTBstat/
	HTBstat.py	-- HTBstat module.
	__init__.py	-- __init__ for the module
	STATpic.py	-- this is a "picture module", contains a class
			   which instantiates "HTB pictures" objects.
py-htbstat-0.2.1.spec	-- attempt to provide you with a spec for rpm
			   building (I used ALT Linux,
			   http://www.altlinux.org/ when have created
			   py-htbstat -- now this spec may be obsolete
			   already)



Installation
============

Place files to appropriate locations :-)

RPM install should work.



Where to use
============

This "scripts set" was designed to be used on linux bridges (I like linux
bridges and HTB on linux bridges:-) and you'll understand this having read
configuration files.

Configuration suppose you may have one or two interfaces with HTB classes
hierarchies configured: one interface is a common situation on routers,
configuration with two interfaces designed for bridge or router with
symmetrical HTB hierarchies -- in this case htbstat.py will draw you both
"incoming" and "outgoing" traffic on one picture.

And it is not of means to be used on boxes with two interfaces with different
hierarchies -- in this case you'll better create two configuration files sets
and scripts sets.



What to read further
====================

Be sure to read carefully configuration files -- there's a lot of
explanations. htbstat.conf then htbstat.cgi. And I tried to comment all
scripts and module files (however this work is not finished yet).



Known issues
============

Sometimes you may refresh browser page right at the moment when rrd bases are
being updated -- you'll see error message. Refresh the page again.



TODO
====

Would be nice to add to STATpic a possibility to summarize counters -- to
see sum of clients rates at one graph. Some improvements may be added to
HTBstat (STATpic) like specifying picture dimensions in config file and
passing as arguments to STATpic etc. And there can be a lot of analyzing
enhancements... Any other ideas, please?



Author(s)
=========

Dmytro O. Redchuk <dor .at. ldc .dot. net>, with a great help of all
VOLZ/LDC team.

Please, do not blame me to much -- this is my the first Python code piece :-)

Actually, I have been working (and I am working further) on another code --
HTB configurator -- and these monitoring-only scripts are much like side
effect :)



CHANGELOG
* Wed Jan 16 2008 Dmytro O. Redchuk <dor@ldc.net> 0.2.3-dor2
- Added consolidation function selection field to cgi script.
- Minor changes: renamed "irate" and "iceil" ( to "rate" and "ceil")
  on pictures and cleaned the code a bit.

* Tue Jan 15 2008 Dmytro O. Redchuk <dor@ldc.net> 0.2.3-dor1
- Fixed `re's in HTBstat.py to accept `backlog [0-9]b [0-9]p' format
  (takes into account packets only though).
- Fixed `kpps' input field interpretation.

* Mon Oct 09 2006 Dmytro O. Redchuk <dor@ldc.net> 0.2.2-dor1
- Changed all (class and scripts) to deal with bits, not kbits
- Fixed `re's in HTBstat.py to accept different `tc -s class' output formats
- rate and ceil values were stored in kbits not bits -- fixed
- Some other fixes for bugs in one-device mode

* Wed Sep 27 2006 Dmytro O. Redchuk <dor@ldc.net> 0.2.1-dor3
- Made it possible to use xxxx:xxxx classids.

* Fri Dec 16 2005 Dmytro O. Redchuk <dor@ldc.net> 0.2-dor1
- Changed method of setting graph upper limit
  (was: rrdtool's LIMIT, now: -u [-r])
- Added wwwpicpath to cgi config file.
- Added htbstatpic.py for getting single picture (png format).

* Thu Mar 24 2005 Dmytro O. Redchuk <dor@ldc.net> 0.1-dor1
- Initial build.



HTBstat/HTBstat.py
#
# $Id: HTBstat.py,v 1.6 2008/01/15 14:57:30 dima Exp $
#
#
# Copyright (c) 2005 Dmytro O. Redchuk
# Copyright (c) 2005 VOLZ Llc
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

import re

from STATpic import STATpic

class HTBstat:
    """HTB class, which contains all the needed HTB attributes
    like rate, ceil, parent, quantum etc."""

    __version__ = "$Revision: 1.6 $"
    # $Source: /var/local/cvs/py-htbstat/HTBstat/HTBstat.py,v $

    def __init__(self, line):
        """Creates HTB class from an input line,
        which should be taken from `tc class show dev <dev>' command."""

        line = line.lower()

        #
        # these may be used to analyze HTB classes hierarchy:
        self.__kidsrate = 0
        self.__kidsceil = 0

        ro = re.compile('root')
        cl = re.compile('class ... ([0-9a-f]+:[0-9a-f]*) ')
        pa = re.compile('parent ([0-9a-f]+:[0-9a-f]*) ')
        ra = re.compile('rate ([0-9]+)([km]{0,1})(bps|bit)')
        ce = re.compile('ceil ([0-9]+)([km]{0,1})(bps|bit)')
        by = re.compile('sent ([0-9]+) bytes')
        pk = re.compile('bytes ([0-9]+) pkts*')
        dr = re.compile('dropped ([0-9]+), ')
        ol = re.compile('overlimits ([0-9]+)')
        le = re.compile('lended: ([0-9]+)')
        bo = re.compile('borrowed: ([0-9]+)')
        to = re.compile('tokens: ([-]*[0-9]+)')
        ct = re.compile('ctokens: ([-]*[0-9]+)')
        bu = re.compile('burst ([0-9]+)([k]*b)')
        cb = re.compile('cburst ([0-9]+)([k]*b)')
        ba = re.compile('backlog ([0-9]+b)* ([0-9])+p')

        # will raise and exception when forced to init
        # with malformed `line':
        try:
            # self.__parent:
            root = ro.search(line)
            parn = pa.search(line)
            if not root:
                self.__parent = parn.group(1)
            else:
                # self.__parent = root.group(0)
                self.__parent = 0
            # self.__class:
            clas = cl.search(line)
            self.__class = clas.group(1)
            #
            # self.__rate:
            rate = ra.search(line)
            if rate:
                self.__rate = int(rate.group(1))
                mult = rate.group(2)
                suff = rate.group(3)
                #
                if suff == 'bit':
                    if mult == 'k':
                        self.__rate *= 1024
                    elif mult == 'm':
                        self.__rate *= 1048576  # 1024 * 1024
                #
                elif suff == 'bps':
                    self.__rate *= 8
                    if mult == 'k':
                        self.__rate *= 1024
                    elif mult == 'm':
                        self.__rate *= 1048576  # 1024 * 1024
            else:
                self.__rate = 0
            #
            # self.__ceil:
            ceil = ce.search(line)
            if ceil:
                self.__ceil = int(ceil.group(1))
                mult = ceil.group(2)
                suff = ceil.group(3)
            else:
                self.__ceil = 0
            #
            if rate > 0:
                if suff == 'bit':
                    if mult == 'k':
                        self.__ceil *= 1024
                    elif mult == 'm':
                        self.__ceil *= 1048576  # 1024 * 1024
                #
                elif suff == 'bps':
                    self.__ceil *= 8
                    if mult == 'k':
                        self.__ceil *= 1024
                    elif mult == 'm':
                        self.__ceil *= 1048576  # 1024 * 1024
            #
            # self.__burst:
            burst = bu.search(line)
            if burst:
                suff = burst.group(2)
                if suff == 'kb':
                    self.__burst = 1024 * int(burst.group(1))
                elif suff == 'mb':
                    self.__burst = 1024 * 1024 * int(burst.group(1))
                else:
                    self.__burst = int(burst.group(1))
            else:
                self.__burst = 0
            # self.__cburst:
            cburst = cb.search(line)
            if cburst:
                self.__cburst = int(cburst.group(1))
            else:
                self.__cburst = 0
            # self.__bytes:
            bytes = by.search(line)
            if bytes:
                self.__bytes = int(bytes.group(1))
            else:
                self.__bytes = 0
            # self.__packets:
            packets = pk.search(line)
            if packets:
                self.__packets = int(packets.group(1))
            else:
                self.__packets = 0
            # self.__dropped:
            drop = dr.search(line)
            self.__dropped = int(drop.group(1))
            # self.__overlimits:
            ovlim = ol.search(line)
            self.__overlimits = int(ovlim.group(1))
            # self.__lended:
            lend = le.search(line)
            if lend:
                self.__lended = int(lend.group(1))
            else:
                self.__lended = 0
            # self.__borrowed:
            borr = bo.search(line)
            if borr:
                self.__borrowed = int(borr.group(1))
            else:
                self.__borrowed = 0
            # self.__tokens:
            tok = to.search(line)
            if tok:
                self.__tokens = int(tok.group(1))
            else:
                self.__tokens = 0
            # self.__ctokens:
            ctok = ct.search(line)
            if ctok:
                self.__ctokens = int(ctok.group(1))
            else:
                self.__ctokens = 0
            # self.__backlog:
            blog = ba.search(line)
            if not blog:
                self.__backlog = 0
            else:
                self.__backlog = int(blog.group(2))
            #
        except AttributeError:
            print '\n  ** AttributeError: \
                    malformed data fed to constructor. **'
            print   '  ** (should be output of \
                    `tc -s class show dev <dev>\') **\n'
            raise
        return

    def __cmp__(self, other):
        """HTB classes can be sorted by class id with __cmp__"""
        return cmp(self.__class, other.__class)

    def rcsid(self):
        """Returns RSC id of HTBstat.py source file."""
        return '$Id: HTBstat.py,v 1.6 2008/01/15 14:57:30 dima Exp $'

    def dump(self):
        """Produces an output,
        similar to `tc class show ...' command."""
        dump = 'class htb ' + self.__class
        if self.__parent == 0:
            dump = dump + ' root '
        else:
            dump = dump + ' parent ' + self.__parent

        dump = dump + ' rate ' + str(self.__rate) + 'bit' + \
                ' ceil ' + str(self.__ceil) + 'bit' + \
                ' burst ' + str(self.__burst) + \
                'b/8 cburst ' + str(self.__cburst) + 'b/8' + \
                '\n Sent ' + str(self.__bytes) + ' bytes ' + \
                str(self.__packets) + ' pkts (dropped ' + \
                str(self.__dropped) + ' overlimits ' + \
                str(self.__overlimits) + ')\n' + \
                '   backlog ' + str(self.__backlog) + 'p\n' + \
                ' lended: ' + str(self.__lended) + \
                ' borrowed: ' + str(self.__borrowed) + '\n' + \
                ' tokens: ' + str(self.__tokens) + \
                ' ctokens: ' + str(self.__ctokens)
        return dump

    #
    def clid(self):
        """Returns HTB class id."""
        return self.__class
    def bytes(self):
        """Returns HTB `bytes sent' counter."""
        return self.__bytes
    def packets(self):
        """Returns HTB `packets sent' counter."""
        return self.__packets
    def dropped(self):
        """Returns HTB `packets dropped' counter."""
        return self.__dropped
    def overlimits(self):
        """Returns HTB `overlimits' counter."""
        return self.__overlimits
    def lended(self):
        """Returns HTB `lended' counter."""
        return self.__lended
    def borrowed(self):
        """Returns HTB `borrowed' counter."""
        return self.__borrowed
    def tokens(self):
        """Returns HTB `tokens' counter."""
        return self.__tokens
    def ctokens(self):
        """Returns HTB `ctokens' counter."""
        return self.__ctokens
    def parent(self):
        """Returns HTB class' parent id."""
        return self.__parent
    def rate(self):
        """Returns HTB class rate value."""
        return self.__rate
    def ceil(self):
        """Returns HTB class ceil value."""
        return self.__ceil
    def burst(self):
        """Returns HTB class burst value."""
        return self.__burst
    def cburst(self):
        """Returns HTB class cburst value."""
        return self.__cburst
    def backlog(self):
        """Returns HTB class backlog value."""
        return self.__backlog
    def kidsrate(self, rate=None):
        """Returns sum of children's rates.
        May be used to analyze HTB classes hierarchy."""
        if rate:
            self.__kidsrate = rate
        return self.__kidsrate
    def kidsceil(self, ceil=None):
        """Returns sum of children's ceils.
        May be used to analyze HTB classes hierarchy."""
        if ceil:
            self.__kidsceil = ceil
        return self.__kidsceil

##############################################################


HTBstat/__init__.py
#
# $Id: __init__.py,v 1.1.1.1 2005/12/16 09:07:40 dima Exp $
#
# Copyright (c) 2005 Dmytro O. Redchuk
# Copyright (c) 2005 VOLZ Llc
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

from HTBstat import *


cgi-bin/htbstat.py
#!/usr/bin/python
#
# $Id: htbstat.py,v 1.11 2008/01/24 09:11:36 dima Exp $
# vim:set ts=4 sw=4 sts=4 expandtab:
#
#
# Copyright (c) 2005 Dmytro O. Redchuk
# Copyright (c) 2005 VOLZ Llc
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

import cgi

import rrdtool
import os
import re
import glob

import pickle
import md5

import time

import HTBstat

# config file:
execfile('/etc/htbstat/htbstat.cgi')

remaddr = os.getenv("REMOTE_ADDR")

######################################################################

print "Content-Type: text/html"     # HTML is following
print                               # blank line, end of headers

print '<html>\n<head>\n<title>%s</title>' % pagetitle
print """
 <style type="text/css">
    pre    {
        text-align: left;
    }
    span.term {
        font-family: 'Fixed', 'Courier New';
    }
    span.rcsid {
        font-family: 'Fixed', 'Courier New';
        font-size: 70%%;
    }
 </style>
 <link rel="stylesheet" type="text/css" href="style.css">
</head>
<body>
"""

######################################################################

keys = { 'parent': 'Parent is ',
     'classid': 'Class ID is ' }

units = { 'packets': 'Packets',
      'bytes': 'Bytes',
      'tokens': 'Tokens' }

timeranges = { '0014400': 'last 2 hours',
        '0118800': 'last 33 hours',
        '0831600': 'last 9 days',
        '3564000': 'last 41 days' }

uppers = { '0.5ceil':  'ceil x .5',
    '1.0ceil':  'ceil',
    '1.1ceil': 'ceil x 1.1',
    '2ceil':  'ceil x 2',
    '4ceil':  'ceil x 4',
    'tenceil': 'ceil x 10',
    'unlim':  'unlim' }

CFs = {
    'MIN': 'MIN',
    'MAX': 'MAX',
    'AVERAGE': 'AVG'
}

######################################################################

# remove old pictures:
oldpics = glob.glob ( '%s/*%s*.png' % ( picpath, remaddr ) )
for oldpic in oldpics:
    os.remove(oldpic)

#
# Make a lists of rrdbases:
#
if devone != disabled:
    irrdlist = glob.glob('%s/*.rrd' % rrdONEpath)
    irrdhash = {}
    irrdcomment = {}

    for it in irrdlist:
        namepat = re.search('([a-f0-9]+:[a-f0-9]+).rrd', it)
        if namepat:
            namestr = namepat.group(1)

            if os.path.exists('/etc/htb'):
                # Connect to htbinit to get comments for the graphs
                htbinitpat = re.search('(1:([0-9]{2,}))', namestr)
                if htbinitpat:
                    tcclass = htbinitpat.group(2)
                    htbinitlist = glob.glob('/etc/htb/%s-*:%s*' % (devone, tcclass))
                    if len(htbinitlist) == 1:
                        commentpat = re.search('[0-9]+:[0-9]+:%s\.(.*)' % tcclass, htbinitlist[0])
                        htbcomment = commentpat.group(1)

                        irrdcomment[namestr] = htbcomment
        else:
            namestr = None

        irrdhash[namestr] = namestr

    iclfiles = glob.glob('%s/*.pickle' % rrdONEpath)
    iclasses = []

    for cls in iclfiles:
        classfile = open(cls, 'r')
        iclass = pickle.load(classfile)
        iclasses.append(iclass)

# the same for devout:
if devtwo != disabled:
    orrdlist = glob.glob('%s/*.rrd' % rrdTWOpath)
    orrdhash = {}
    
    for it in orrdlist:
        namepat = re.search('([a-f0-9]+:[a-f0-9]+).rrd', it)
        namestr = namepat.group(1)
        orrdhash[namestr] = namestr
    
    oclfiles = glob.glob('%s/*.pickle' % rrdTWOpath)
    oclasses = []
    
    for cls in oclfiles:
        classfile = open(cls, 'r')
        oclass = pickle.load(classfile)
        oclasses.append(oclass)

#
# list of classes
# and loading classes:

######################################################################

#
# twidding with cgi:
#
F = cgi.FieldStorage()

#
# too long :-)
F.gv = F.getvalue
F.gf = F.getfirst


# get CGI values from the form:
dev, iclid, oclid, ikey, okey, unit, limits, trange, upper, \
                pkts_upperlim, pkts_forcelim, ceil_upperlim, ceil_forcelim, \
                CF = \
    F.gv("dev"), F.gf("iclid"), F.gv("oclid"), \
    F.gv("ikey"), F.gv("okey"), F.gf("unit"), \
    F.gv("limits"), F.gf("trange"), F.gf("upper"), \
    F.gv("pupperlim"), F.gv("pforcelim"), F.gv("cupperlim"), \
    F.gv("cforcelim"), F.gv("CF")


if dev not in (devone, devtwo):
    if devone != disabled:
        dev = devone
    else:
        dev = devtwo

if not iclid:
    iclid = [ defclid ]
else:
    iclid = F.getlist("iclid")

if not oclid:
    oclid = [ defclid ]
else:
    oclid = F.getlist("oclid")

if ikey not in keys: ikey = 'classid'
if okey not in keys: okey = 'classid'

if dev == devone:
    clid = iclid
    key = ikey
else:
    clid = oclid
    key = okey

if not unit:
    unit = [ 'bytes' ]
else:
    unit = F.getlist("unit")

if not trange:
    trange = [ '0118800' ]
else:
    trange = F.getlist("trange")


if not upper:
    upper = [ 'ceil' ]
else:
    upper = F.getlist("upper")


if not ceil_upperlim:
    ceil_upperlim = 1
else:
    ceil_upperlim = float(ceil_upperlim)

if not ceil_forcelim:
    ceil_forcelim = ''
else:
    ceil_forcelim = 'checked'

if not pkts_upperlim:
    pkts_upperlim = 1000.0
else:
    pkts_upperlim = float(pkts_upperlim) * 1000.0

if not pkts_forcelim:
    pkts_forcelim = ''
else:
    pkts_forcelim = 'checked'


if CF not in CFs.keys():
    CF = 'AVERAGE'

#
# all cgi values are set.
#
######################################################################

#
# Here we need to create a list with clids to be drawn:
#

if dev == devone:
    classfilelist = iclasses
else:
    classfilelist = oclasses
    # swap pathes:
    (rrdONEpath, rrdTWOpath) = (rrdTWOpath, rrdONEpath)

#

# all the classes (possibly to process) for this flow direction:
classfilelist.sort()

#
#
statpic = HTBstat.STATpic()
statpic.picpath(picpath)
statpic.units(unit)

# we need this only to `dump()':
htbclasslist = []

#
# make list of classes to be `pictured':
if key == 'classid':
    for cl in clid:
        htb = getbyclid(classfilelist, cl)
        htbclasslist.append(htb)
        if devone != disabled:
            rrdONEfile = '%s/%s.rrd' % ( rrdONEpath, htb.clid() )
        else:
            rrdONEfile = None
        if devtwo != disabled:
            rrdTWOfile = '%s/%s.rrd' % ( rrdTWOpath, htb.clid() )
        else:
            rrdTWOfile = None
        statpic.addhtbclass(htb, rrdONEfile, rrdTWOfile)
elif key == 'parent':
    for htb in classfilelist:
        for cl in clid:
            if htb.parent() == cl:
                htbclasslist.append(htb)
                if devone != disabled:
                    rrdONEfile = '%s/%s.rrd' % ( rrdONEpath, htb.clid() )
                else:
                    rrdONEfile = None
                if devtwo != disabled:
                    rrdTWOfile = '%s/%s.rrd' % ( rrdTWOpath, htb.clid() )
                else:
                    rrdTWOfile = None
                statpic.addhtbclass(htb, rrdONEfile, rrdTWOfile, irrdcomment[htb.clid()])
else:
    for cl in clid:
        htb = getbyclid(classfilelist, cl)
        htbclasslist.append(htb)
        if devone != disabled:
            rrdONEfile = '%s/%s.rrd' % ( rrdONEpath, htb.clid() )
        else:
            rrdONEfile = None
        if devtwo != disabled:
            rrdTWOfile = '%s/%s.rrd' % ( rrdTWOpath, htb.clid() )
        else:
            rrdTWOfile = None
        statpic.addhtbclass(htb, rrdONEfile, rrdTWOfile)


#
#
# HTML page.
#
#

print """
<div id="body">
    <h1>%s</h1>
    <form>
    <table class="form"><tr>
        <th>Select source:</th>
        <th>What to display:</th>
    </tr><tr>
        <td>
            <table class="options">
""" % pageheader


# if both are enabled:
if disabled not in (devone, devtwo):
    if dev == devone:
        print "<tr>"
        print """<th colspan="2">
        <input type="radio" name="dev" value="%s" checked>%s
        </th>""" % (devone, devonename)

        if devtwo != disabled:
            print """<th colspan="2">
            <input type="radio" name="dev" value="%s">%s
            </th>""" % (devtwo, devtwoname)

        print "</tr>"

    elif dev == devtwo:
        print "<tr>"
        if devone != disabled:
            print """<th colspan="2">
            <input type="radio" name="dev" value="%s">%s
            </th>""" % (devone, devonename)

        print """<th colspan="2">
            <input type="radio" name="dev" value="%s" checked>%s
            </th>""" % (devtwo, devtwoname)

        print "</tr>"
else:
    print "<tr><th colspan='2'>"
    if dev == devone:
        print devonename
    else:
        print devtwoname

print '</th></tr>'


if devone != disabled:
    print '<TD>%s' % getSelect(keys, 'ikey', ikey)
    print '<TD>%s' % getSelect(irrdhash, 'iclid', clid, 1, 6, \
                        ' Use Control key to select multiple classes. ')

if devtwo != disabled:
    print '<TD>%s' % getSelect(keys, 'okey', okey)
    print '<TD>%s' % getSelect(orrdhash, 'oclid', clid, 1, 6, \
                        ' Use Control key to select multiple classes. ')

print '</TABLE>'

print '<TD align="center"><table class="options">'

print '<TR><TD>%s' % getSelect(units, 'unit', unit, 1, 3, \
                        ' Use Control key to select multiple entries. ')

print '<TD>%s' % getSelect(timeranges, 'trange', trange, 1, 4, \
                        ' Use Control key to select multiple timeranges. ')

print """
</TABLE>
"""

if limits == 'on':
    print '<BR><INPUT type="checkbox" name="limits" checked>',
else:
    print '<BR><INPUT type="checkbox" name="limits">',

print """
&nbsp;Show <SPAN class="term">rate</SPAN>
and <SPAN class="term">ceil</SPAN> values &nbsp;
<TR><TD align="right">Upper limit:<TT>
    <INPUT type="text" size="3" maxsize="3" name="cupperlim"
        value="%s" align="right">&nbsp;x&nbsp;ceil
""" % ceil_upperlim


print """
    (<INPUT type="checkbox" name="cforcelim" value="forced" %s>
        force hard)<BR>
""" % ceil_forcelim

print """
    <INPUT type="text" size="3" maxsize="3" name="pupperlim"
        value="%s" align="right">&nbsp; &nbsp;kpps
""" %  str(pkts_upperlim/1000)

print """
    (<INPUT type="checkbox" name="pforcelim" value="forced" %s>
        force hard)</TT>
""" % pkts_forcelim


print '<TD align="center">CF: %s' % \
            getSelect(CFs, 'CF', CF, 0, 0, "rrd's Consolidation Function")


print """
    &nbsp; <INPUT type="submit" value="Submit">

</FORM>
</TABLE>
<TR><TD align="center">
"""

# if we asked to draw rate and ceil lines
# (`limits' should be `on' or None ideally):
statpic.needlimits(limits)

# set upper limit for bytes:
statpic.ceil_upperlim(ceil_upperlim)

# force hard upper limit, if needed:
if ceil_forcelim == 'checked':
    statpic.ceil_forcelim(1)
else:
    statpic.ceil_forcelim(0)


# set upper limit for packets:
statpic.pkts_upperlim(pkts_upperlim)

# force hard upper limit, if needed:
if pkts_forcelim == 'checked':
    statpic.pkts_forcelim(1)
else:
    statpic.pkts_forcelim(0)

# set CF:
statpic.CF(CF)

# add timeranges:
for i in range(len(trange)):
    if i == 0:
        statpic.timerange(int(trange[i]))
    else:
        statpic.addtimerange(int(trange[i]))


# generate a picture ("base-") name:
md5key  = md5.new(str(time.time())).hexdigest()
picname = '%s-%s.png' % ( remaddr, md5key ) 

# draw all the pictures!
piclist = statpic.draw(picname)

# print HTML links to images:
if len(piclist) == 0:
    print '<CENTER><B>Nothing to draw.</B></CENTER>'
else:
    for pic in piclist:
        print '<IMG src="%s/%s">' % ( wwwpicpath, pic )

print '<br style="clear: both" />'
# dump class only `if':
if key == 'classid' or not key:
    print '<TR><TD align="center">'
    for htbclass in htbclasslist:
        print """
            <PRE style="border: solid thin gray; padding: 3 3 3 3;">
%s                      <A href="dumpclass.py?dev=%s&clid=%s"
                            target="_blank">Monitor this class</A>
</PRE>
""" % ( htbclass.dump(), dev, htbclass.clid() )

print """
</TABLE>
<HR>
<TABLE border="0" width="90%%" align="center">
<TR><TD>
<SPAN class="rcsid">Copyright (c) 2005&ndash;2008 <A href="mailto:%s@%s.%s">Dmytro O. Redchuk</A></SPAN><BR>
<SPAN class="rcsid">Copyright (c) 2005&ndash;2008 <A href="http://www.volz.ua/">VOLZ Llc</A></SPAN><BR>
""" % ('brownian.box', 'gmail', 'com')

# to get RCS Id:-)
sample = iclasses[0]

#print """
#<TD align="right">
#    <SPAN class="rcsid">htbstat.py: %s</SPAN><BR>
#    <SPAN class="rcsid">HTBstat: %s</SPAN><BR>
#    <SPAN class="rcsid">STATpic: %s</SPAN><BR>
#"""
# % ( rcsid, sample.rcsid(), statpic.rcsid() )
print '</TABLE>'

# WOW! All is well.
